toucan_connectors/clickhouse/clickhouse_connector.py

```diff
@@ -0,0 +1,117 @@
+    query: constr(min_length=1) = Field(
+        None,
+        description='You can write a custom query against your '
+        'database here. It will take precedence over '
+        'the "table" parameter above',
+        widget='sql',
+    )
+    table: constr(min_length=1) = Field(
+        None,
+        description='The name of the data table that you want to '
+        'get (equivalent to "SELECT * FROM '
+        'your_table")',
+    )
+
+    class Config:
+        @staticmethod
+        def schema_extra(schema: Dict[str, Any], model: Type['ClickhouseDataSource']) -> None:
+            keys = schema['properties'].keys()
+            prio_keys = [
+                'database',
+                'table',
+                'query',
+                'parameters',
+            ]
+            new_keys = prio_keys + [k for k in keys if k not in prio_keys]
+            schema['properties'] = {k: schema['properties'][k] for k in new_keys}
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        query = data.get('query')
+        table = data.get('table')
+        if query is None and table is None:
+            raise ValueError("'query' or 'table' must be set")
+        elif query is None and table is not None:
+            self.query = f'select * from {table};'
+
+    @classmethod
+    def get_form(cls, connector: 'ClickhouseConnector', current_config):
+        """
+        Method to retrieve the form with a current config
+        For example, once the connector is set,
+        - we are able to give suggestions for the `database` field
+        - if `database` is set, we are able to give suggestions for the `table` field
+        """
+        constraints = {}
+
+        with suppress(Exception):
+            connection = clickhouse_driver.connect(
+                f'clickhouse://{connector.user}:{connector.password.get_secret_value() if connector.password else ""}@{connector.host}:{connector.port}'
+            )
+
+            # Always add the suggestions for the available databases
+
+            with connection.cursor() as cursor:
+                cursor.execute('SHOW DATABASES')
+                res = cursor.fetchall()
+                available_dbs = [db_name for (db_name,) in res if db_name != 'system']
+                constraints['database'] = strlist_to_enum('database', available_dbs)
+
+                if 'database' in current_config:
+                    cursor.execute(
+                        f"""SELECT name FROM system.tables WHERE database = '{current_config["database"]}'"""
+                    )
+                    res = cursor.fetchall()
+                    available_tables = [table[0] for table in res]
+                    constraints['table'] = strlist_to_enum('table', available_tables)
+
+        return create_model('FormSchema', **constraints, __base__=cls).schema()
+
+
+class ClickhouseConnector(ToucanConnector):
```

Comment:
@raphaelvignes I think we should add a configuration option to support SSL wrapped TCP connection ("secure connection"). The protocol part of the url is not `clickhouse` in this case but `clickhouses` (with a S)

https://clickhouse-driver.readthedocs.io/en/latest/api.html#clickhouse_driver.Client.from_url

Reply:
Thanks, I added this option.